### PR TITLE
Avoid blocking during concurrent hot threads calls

### PR DESCRIPTION
**Pull request by DaveCTurner**
Today we avoid concurrent hot threads detections using a mutex, but the
user controls how long each detection takes. An abusive call could block
lots of other threads for a very long time. This commit changes the
behaviour to bail out early with an informative message if there is
already a detection in progress.

### CODE REVIEW

File: server/src/test/java/org/elasticsearch/monitor/jvm/HotThreadsTests.java [hunk 1]
```diff
@@ -939,6 +942,55 @@ public void testInnerDetectCPUModeTransportThreads() throws Exception {
...
+            }
+        });
+        backgroundThread.start();
+
+        safeAwait(barrier);
+        try (var stringWriter = new StringWriter()) {
+            new HotThreads().detect(stringWriter);
+            assertEquals(ALREADY_RUNNING_MESSAGE, stringWriter.toString());
+        } catch (Exception e) {
+            fail(e);
```
Comment (thecoop) [inline server/src/test/java/org/elasticsearch/monitor/jvm/HotThreadsTests.java:987]: Why not just let the exception bubble up?

File: server/src/main/java/org/elasticsearch/monitor/jvm/HotThreads.java [hunk 2]
```diff
@@ -35,12 +35,14 @@
...
 import java.util.Map;
+import java.util.concurrent.Semaphore;
 import java.util.concurrent.TimeUnit;
 import java.util.function.ToLongFunction;
 
 public class HotThreads {
 
-    private static final Object mutex = new Object();
+    private static final Semaphore permits = new Semaphore(1);
+    static final String ALREADY_RUNNING_MESSAGE = "hot threads already running";
```
Comment (thecoop) [inline server/src/main/java/org/elasticsearch/monitor/jvm/HotThreads.java:45]: Maybe the message should indicate more that this functionality is being used in the wrong way

Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/monitor/jvm/HotThreads.java]: How about 46add2957df0b31a691aadeb58ea8d9b4c965c4d?

**Comment by DaveCTurner** [comment]
I am confused by the failure in https://gradle-enterprise.elastic.co/s/2t27ysy72qqlq/console-log/task/:server:internalClusterTest, I expected that we would only be running tests sequentially, but these tests failed precisely because of this change.

File: server/src/main/java/org/elasticsearch/monitor/jvm/HotThreads.java (hunk 2 above)
Comment (thecoop) [reply server/src/main/java/org/elasticsearch/monitor/jvm/HotThreads.java]: Yeah, much clearer

**Review by thecoop — APPROVED** [review]

**Comment by DaveCTurner** [comment]
Eh I see now, this isn't going to work as-is in an `ESIntegTestCase` because there we do fan out these requests across multiple nodes, but they're all in the same JVM, so a static Semaphore is no good. 😞

**Pull request closed without merging**